### PR TITLE
Add googletest test framework

### PR DESCRIPTION
**Pull request by oliverlee**
- [x] merge #106
- [x] determine reason for test failures

### CODE REVIEW

**Comment by oliverlee** [comment]
CI fails as we cache submodules (but we just added a new one and it's empty).

**Comment by oliverlee** [comment]
Is COBS decode being called correctly in the random data test? I get the follow errors when I run the test:
```
oliver@canopus:~/repos/phobos/tests/build$ ./test_cobs_random
Running main() from gtest_main.cc
[==========] Running 10 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 10 tests from CobsRandomInput/CobsRandomDataTest
[ RUN      ] CobsRandomInput/CobsRandomDataTest.RandomInput/0
/Users/oliver/repos/phobos/tests/test_cobs_random.cc:47: Failure
      Expected: des_res.status
      Which is: 4-byte object <03-00 00-00>
To be equal to: cobs::DecodeResult::Status::OK
      Which is: 4-byte object <00-00 00-00>
[  FAILED  ] CobsRandomInput/CobsRandomDataTest.RandomInput/0, where GetParam() = 10 (0 ms)
[ RUN      ] CobsRandomInput/CobsRandomDataTest.RandomInput/1
/Users/oliver/repos/phobos/tests/test_cobs_random.cc:47: Failure
      Expected: des_res.status
      Which is: 4-byte object <03-00 00-00>
To be equal to: cobs::DecodeResult::Status::OK
      Which is: 4-byte object <00-00 00-00>
[  FAILED  ] CobsRandomInput/CobsRandomDataTest.RandomInput/1, where GetParam() = 100 (1 ms)
[ RUN      ] CobsRandomInput/CobsRandomDataTest.RandomInput/2
/Users/oliver/repos/phobos/tests/test_cobs_random.cc:47: Failure
      Expected: des_res.status
      Which is: 4-byte object <03-00 00-00>
To be equal to: cobs::DecodeResult::Status::OK
      Which is: 4-byte object <00-00 00-00>
[  FAILED  ] CobsRandomInput/CobsRandomDataTest.RandomInput/2, where GetParam() = 200 (0 ms)
...
```

File: tests/test_cobs.cc [hunk 1]
```diff
@@ -83,25 +66,43 @@ void test_encode_decode(std::vector<Rep> decoded, std::vector<Rep> encoded) {
...
+    EXPECT_EQ(dec_act_res.written, dec_exp_len);
+    test_equal_buffers(dec_exp, sizeof(dec_exp), dec_act, sizeof(dec_act));
 }
 
-int main() {
-    test_max_encoded_length();
-    test_max_decoded_length();
+} // namespace
+
+TEST(CobsLengthCalculation, MaxEncodeLength) {
```
Comment (mickvangelderen) [inline tests/test_cobs.cc]: Assuming the first argument is the test namespace, I think we should keep the namespace of the test equal to the namespace of the code under test, so in this case "cobs".

File: tests/test_cobs_random.cc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+void CobsRandomDataTest::test_encode_decode() {
+    // encode b1 to b2
+    const cobs::EncodeResult enc_res = cobs::encode(b1.data(), b1.size(), b2.data(), b2.capacity());
+    ASSERT_EQ(enc_res.status, cobs::EncodeResult::Status::OK);
+    b2.resize(enc_res.written);
+
+    // decode b2 to b3
+    const cobs::DecodeResult des_res = cobs::decode(b2.data(), b2.size(), b3.data(), b3.capacity());
```
Comment (mickvangelderen) [inline tests/test_cobs_random.cc]: `des_res` should probably be `dec_res`

File: tests/test_cobs_random.cc [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    ASSERT_EQ(des_res.status, cobs::DecodeResult::Status::OK);
+    b3.resize(des_res.written);
```
Comment (mickvangelderen) [inline tests/test_cobs_random.cc]: also assert `ASSERT_EQ(des_res.read, b2.size())`

File: tests/test_cobs_random.cc [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    // test b1 == b3
+    ASSERT_EQ(b1.size(), b3.size());
```
Comment (mickvangelderen) [inline tests/test_cobs_random.cc]: Use `test_equal_buffer` from the other test file.

**Comment by mickvangelderen** [comment]
@oliverlee Random test code code seems correct. I'd have to inspect the input data, encoded (input  data) and decoded (encoded (input data)).

File: tests/test_cobs.cc (hunk 1 above)
Comment (oliverlee) [inline tests/test_cobs.cc]: fixed

File: tests/test_cobs_random.cc (hunk 2 above)
Comment (oliverlee) [inline tests/test_cobs_random.cc]: fixed

File: tests/test_cobs_random.cc (hunk 3 above)
Comment (oliverlee) [inline tests/test_cobs_random.cc]: fixed

File: tests/test_cobs_random.cc (hunk 4 above)
Comment (oliverlee) [inline tests/test_cobs_random.cc]: Fixed. `test_equal_buffer` has been declared and defined in files `test_cobs_util.cc` and `test_cobs_util.h`.

File: tests/test_cobs.cc [hunk 5]
```diff
@@ -83,25 +58,43 @@ void test_encode_decode(std::vector<Rep> decoded, std::vector<Rep> encoded) {
...
+    EXPECT_EQ(dec_act_res.written, dec_exp_len);
+    test_equal_buffers(dec_exp, sizeof(dec_exp), dec_act, sizeof(dec_act));
 }
 
-int main() {
-    test_max_encoded_length();
-    test_max_decoded_length();
+} // namespace
+
+TEST(Cobs, MaxEncodeLength) {
```
Comment (mickvangelderen) [inline tests/test_cobs.cc]: Is it possible to use snake_case? I find it more readable in general and we use it for functions which is what we are testing.

Comment (oliverlee) [inline tests/test_cobs.cc]: I don't mind. I was following the examples in the gtest primer but I also prefer snake_case.

File: tests/test_cobs.cc [hunk 6]
```diff
@@ -83,25 +58,43 @@ void test_encode_decode(std::vector<Rep> decoded, std::vector<Rep> encoded) {
...
+    EXPECT_EQ(dec_act_res.written, dec_exp_len);
+    test_equal_buffers(dec_exp, sizeof(dec_exp), dec_act, sizeof(dec_act));
 }
 
-int main() {
-    test_max_encoded_length();
-    test_max_decoded_length();
+} // namespace
+
+TEST(Cobs, max_encode_length) {
```
Comment (mickvangelderen) [inline tests/test_cobs.cc]: `Cobs` -> `cobs`

**Comment by oliverlee** [comment]
Looks good.

**Comment by oliverlee** [comment]
Are the tests running? It doesn't look like it from the logs.

**Pull request merged**